omega/DecisionTree/DecisionTreeClassifier.py

```diff
@@ -0,0 +1,230 @@
+        else:
+            left = (torch.DoubleTensor(), torch.DoubleTensor())
+
+        mask = torch.nonzero(1 - torch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            right = (torch.index_select(x, 0, mask), torch.index_select(y, 0, mask))
+        else:
+            right = (torch.DoubleTensor(), torch.DoubleTensor())
+        return left, right
+
+    def fit(self, x, y):
+        """Trains the classifier on the training data and forms the decision tree
+
+        Parameters
+        ----------
+        x : training data
+        y : training labels
+
+        Returns
+        -------
+        A dictionary containing the tree.
+        """
+        if isinstance(x, np.ndarray):
+            x = torch.from_numpy(x)
+        if isinstance(y, np.ndarray):
+            y = torch.from_numpy(y)
+        if self.__use_cuda:
+            x = x.cuda()
+            y = y.cuda()
+        self.__root = self.__build_tree(x, y)
+        return self.__root
+
+    def __build_tree(self, x, y):
+        """Initiates the tree building process by calling for the first split.
+
+        Parameters
+        ----------
+        x : training data
+        y : training labels
+
+        Returns
+        -------
+        A dictionary containing the tree.
+        """
+        root = self.__get_split(x, y)
+        self.__split(root, 1)
+        return root
+
+    def __get_split(self, x, y):
+        """Finds the best split given the training data and labels
+
+        Parameters
+        ----------
+        x : training data
+        y : training labels
+
+        Returns
+        -------
+        A node for the tree with best split value given the data.
+        """
+        b_index, b_value, b_score, b_groups = 999, 999, 999, None
+
+        classes = list()
+        for val in y:
+            if val not in classes:
+                classes.append(val)
+
+        for index in range(x.size()[1]):
+            for row in x:
+                groups = self.__test_split(index, row[index], x, y)
+                score = self.__gini_index(groups[0], groups[1], classes)
+                if score < b_score:
+                    b_index, b_value, b_score, b_groups = index, row[index], score, groups
+        return {'index': b_index, 'value': b_value, 'groups': b_groups}
+
+    def __to_terminal(self, group):
+        """Determines the label of a terminal node.
+
+        Parameters
+        ----------
+        group : list
+            list contains two elements, data and labels of the rows associated with the split.
+
+        Returns
+        -------
+        target label of a terminal node.
+        """
+        res = torch.mode(group[1])[0][0]
+        return res
+
+    def __split(self, node, depth):
+        """Build the subtree given a node, keeps track of maximum depth of the tree.
+
+        Parameters
+        ----------
+        node : node for which the subtree is built
```

Comment:
Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html